A51-custom-backend-metrics.md

```diff
@@ -380,21 +415,122 @@ public interface OrcaOobReportListener {
+metrics. They can obtain a reference to the `CallMetricRecorder` from the server
+context. All the methods are thread safe.
 
-TODO(nicolasnoble): Fill this in.
+```c++
+class CallMetricRecorder {
+ public:
+   ~CallMetricRecorder() = default;
+
+  // Records a call metric measurement for CPU utilization.
+  // Multiple calls to this method will override the stored value.
+  // Values outside of the valid range [0, 1] are ignored.
+  CallMetricRecorder& RecordCpuUtilizationMetric(double value);
+
+  // Records a call metric measurement for memory utilization.
+  // Multiple calls to this method will override the stored value.
+  // Values outside of the valid range [0, 1] are ignored.
+  CallMetricRecorder& RecordMemoryUtilizationMetric(double value);
+
+  // Records a call metric measurement for queries per second.
+  // Multiple calls to this method will override the stored value.
+  // Values outside of the valid range [0, infy) are ignored.
+  CallMetricRecorder& RecordQpsMetric(double value);
+
+  // Records a call metric measurement for utilization.
+  // Multiple calls to this method with the same name will
+  // override the corresponding stored value. The lifetime of the
+  // name string needs to be longer than the lifetime of the RPC
+  // itself, since it's going to be sent as trailers after the RPC
+  // finishes. It is assumed the strings are common names that
+  // are global constants.
+  // Values outside of the valid range [0, 1] are ignored.
+  CallMetricRecorder& RecordUtilizationMetric(string_ref name,
+                                                      double value);
+
+  // Records a call metric measurement for request cost.
+  // Multiple calls to this method with the same name will
+  // override the corresponding stored value. The lifetime of the
+  // name string needs to be longer than the lifetime of the RPC
+  // itself, since it's going to be sent as trailers after the RPC
+  // finishes. It is assumed the strings are common names that
+  // are global constants.
+  CallMetricRecorder& RecordRequestCostMetric(string_ref name,
+                                                      double value);
+};
 
-#### OOB Reporting APIs
+class ServerContextBase {
+ public:
+  // Get the a CallMetricRecorder object for the current RPC.
+  // Use it to record metrics during your RPC to send back to the
+  // client in order to make load balancing decisions. This will
+  // return nullptr if the feature hasn't been enabled using
+  // EnableCallMetricRecording().
+  CallMetricRecorder* GetCallMetricRecorder();
+};
+```
 
-C++ provides the following API for running the ORCA service:
+To enable per-call load reporting, an application must enable call metric recording when building the server.
+The application may optionally pass `ServerMetricRecorder` to also include machine-wide metrics in per-call load reports.
+
+```c++
+std::unique_ptr<ServerMetricRecorder> server_metric_recorder_ =
```

Comment:
No need for the trailing underscore in the variable name here.

Same thing on line 530.

Reply:
Done.